### PR TITLE
Stub for representable profunctor

### PR DESCRIPTION
**Pull request by tetrapharmakon**
Any functor `F : C -> D` induces two "representable" profunctors 

`hom(F,1) : (c,d) \mapsto D(Fc,d)`

and 

`hom(1,F) : (d,c) \mapsto D(d, Fc)`

### CODE REVIEW

File: src/Categories/Functor/Profunctor.agda [hunk 1]
```diff
@@ -4,13 +4,91 @@ module Categories.Functor.Profunctor where
...
+      ; cong = λ x → begin _ ≈⟨ refl⟩∘⟨ x ⟩∘⟨refl ⟩ _ ∎
+      }
+    ; identity = λ {x} {y} {y'} y≈y' → begin
+        F₁ C.id ∘ y ∘ D.id ≈⟨ elimˡ D identity ⟩
+        y ∘ D.id             ≈⟨ D.identityʳ ⟩
+        y                    ≈⟨ y≈y' ⟩
+        y'                   ∎
+    ; homomorphism = λ { {f = f0 , f1} {g = g0 , g1} {x} {y} x≈y → begin
+        F₁ (g1 C.∘ f1) ∘ x ∘ f0 ∘ g0    ≈⟨ refl⟩∘⟨ x≈y ⟩∘⟨refl ⟩
+        F₁ (g1 C.∘ f1) ∘ y ∘ f0 ∘ g0    ≈⟨ Functor.homomorphism F ⟩∘⟨refl ⟩
```
Comment (JacquesCarette) [inline src/Categories/Functor/Profunctor.agda:78]: `F.homomorphism`

File: src/Categories/Functor/Profunctor.agda [hunk 2]
```diff
@@ -4,13 +4,91 @@ module Categories.Functor.Profunctor where
...
+        (F₁ g1 ∘ F₁ f1) ∘ y ∘ f0 ∘ g0 ≈⟨ D.assoc ⟩
+        F₁ g1 ∘ F₁ f1 ∘ y ∘ f0 ∘ g0       ≈⟨ refl⟩∘⟨ D.Equiv.sym assoc² ⟩
```
Comment (JacquesCarette) [inline src/Categories/Functor/Profunctor.agda:80]: these 2 lines are some kind of pull/push.

**Review by JacquesCarette — CHANGES_REQUESTED** [review]
One big thing, and then a few little fixes.

File: src/Categories/Functor/Profunctor.agda [hunk 3]
```diff
@@ -89,19 +91,46 @@ module Profunctor {o ℓ e} {o′} (C : Category o ℓ e) (D : Category o′ ℓ
...
-    ; identityˡ = {!   !}
-    ; identityʳ = {!   !}
-    ; identity² = {!   !}
-    ; equiv = {!   !}
-    ; ∘-resp-≈ = {!   !}
-    }
\ No newline at end of file
+-- each Prof(C,D) is a category
+homProf : {o o′ ℓ e : Level} → (C : Category o ℓ e) → (D : Category o′ ℓ e) → Category _ _ _
+homProf {o} {o′} {ℓ} {e} C D = record
```
Comment (JacquesCarette) [inline src/Categories/Functor/Profunctor.agda:96]: you don't use the levels below, maybe omit binding them needlessly here?

File: src/Categories/Functor/Profunctor.agda [hunk 4]
```diff
@@ -89,19 +91,46 @@ module Profunctor {o ℓ e} {o′} (C : Category o ℓ e) (D : Category o′ ℓ
...
+  { Obj = Profunctor C D
+  ; _⇒_ = λ P Q → NaturalTransformation P Q
```
Comment (JacquesCarette) [inline src/Categories/Functor/Profunctor.agda:98]: does eta contracting this work?

File: src/Categories/Functor/Profunctor.agda [hunk 5]
```diff
@@ -89,19 +91,46 @@ module Profunctor {o ℓ e} {o′} (C : Category o ℓ e) (D : Category o′ ℓ
...
+  ; _∘_ = _∘ᵥ_
+  ; assoc     = λ { {f = f} {g} {h} → assoc-lemma {f = f} {g} {h}}
+  ; sym-assoc = λ { {f = f} {g} {h} → sym-assoc-lemma {f = f} {g} {h}}
+  ; identityˡ = λ { {f = f} → id-lemmaˡ {f = f}}
+  ; identityʳ = λ { {f = f} → id-lemmaʳ {f = f} }
+  ; identity² = λ z → z
+  ; equiv = ≃-isEquivalence
+  ; ∘-resp-≈ = λ { {f = f} {h} {g} {i} eq eq' → ∘ᵥ-resp-≃ {f = f} {h} {g} {i} eq eq' }
+  }
+  where
```
Comment (JacquesCarette) [inline src/Categories/Functor/Profunctor.agda:110]: Can't these 1-liners not be inlined in the above?  As `D` is visible, you could
```
  where
    module D = Category Y
```
and then everything is quite simple.

File: src/Categories/Morphism/Reasoning/Core.agda [hunk 6]
```diff
@@ -48,8 +48,8 @@ module Utils where
   assoc²' : (i ∘ (h ∘ g)) ∘ f ≈ i ∘ (h ∘ (g ∘ f))
   assoc²' = Equiv.trans assoc (∘-resp-≈ʳ assoc)
 
-  assoc²'' : (i ∘ h) ∘ g ∘ f ≈ i ∘ ((h ∘ g) ∘ f)
-  assoc²'' = Equiv.trans assoc (∘-resp-≈ʳ  (Equiv.sym assoc))
+  assoc²'' : i ∘ ((h ∘ g) ∘ f) ≈ (i ∘ h) ∘ (g ∘ f)
+  assoc²'' = Equiv.trans (∘-resp-≈ʳ assoc) (Equiv.sym assoc)
```
Comment (JacquesCarette) [inline src/Categories/Morphism/Reasoning/Core.agda]: May as well use `sym-assoc` instead of `Equiv.sym assoc` ?

**Review by JacquesCarette — CHANGES_REQUESTED** [review]
Nice. Just a few minor tweaks, and this is ready.

File: src/Categories/Functor/Profunctor.agda (hunk 5 above)
Comment (iwilare) [reply src/Categories/Functor/Profunctor.agda:110]: We thought about inlining these lemmas, but we can't seem to get it to work without generalizing those vertical composition properties over any two categories. The implicit arguments on the 1-liners should be:

```agda
  ; assoc     = λ { {f = f} {g} {h} → assoc-lemma     {C = Product (Category.op D) C} {D = Setoids ℓ e} {f = f} {g} {h}}
  ; sym-assoc = λ { {f = f} {g} {h} → sym-assoc-lemma {C = Product (Category.op D) C} {D = Setoids ℓ e} {f = f} {g} {h}}
  ; identityˡ = λ { {f = f} → id-lemmaˡ {C = Product (Category.op D) C} {D = Setoids ℓ e} {f = f} }
  ; identityʳ = λ { {f = f} → id-lemmaʳ {C = Product (Category.op D) C} {D = Setoids ℓ e} {f = f} }
```

but unfortunately directly substituting in the lemmas does not work. If you manage to find some other way to remove them let us know!

Comment (tetrapharmakon) [reply src/Categories/Functor/Profunctor.agda:110]: How did this end?

**Review by JacquesCarette — APPROVED** [review]

File: src/Categories/Functor/Profunctor.agda (hunk 5 above)
Comment (JacquesCarette) [reply src/Categories/Functor/Profunctor.agda:110]: It is 'ending' in me pulling it in as is (once the build passes). It wasn't clear to me that assoc (etc) were being called 'at' such different categories. So the current setup is just as nice as the one with the implicits spelled out.

**Pull request merged**